clients/java/src/main/java/io/camunda/zeebe/client/impl/worker/JobStreamerImpl.java

```diff
@@ -0,0 +1,201 @@
+/*
+ * Copyright © 2017 camunda services GmbH (info@camunda.com)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.camunda.zeebe.client.impl.worker;
+
+import io.camunda.zeebe.client.api.ZeebeFuture;
+import io.camunda.zeebe.client.api.command.FinalCommandStep;
+import io.camunda.zeebe.client.api.command.StreamJobsCommandStep1.StreamJobsCommandStep3;
+import io.camunda.zeebe.client.api.response.ActivatedJob;
+import io.camunda.zeebe.client.api.response.StreamJobsResponse;
+import io.camunda.zeebe.client.api.worker.BackoffSupplier;
+import io.camunda.zeebe.client.api.worker.JobClient;
+import io.camunda.zeebe.client.impl.Loggers;
+import io.grpc.Status;
+import io.grpc.StatusRuntimeException;
+import java.time.Duration;
+import java.util.List;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.Consumer;
+import net.jcip.annotations.GuardedBy;
+import net.jcip.annotations.ThreadSafe;
+import org.slf4j.Logger;
+
+@ThreadSafe
+final class JobStreamerImpl implements JobStreamer {
+  private static final Logger LOGGER = Loggers.JOB_WORKER_LOGGER;
+
+  private final JobClient jobClient;
+  private final String jobType;
+  private final String workerName;
+  private final Duration timeout;
+  private final List<String> fetchVariables;
+  private final Duration requestTimeout;
+  private final BackoffSupplier backoffSupplier;
+  private final ScheduledExecutorService executor;
+  private final Lock streamLock;
```

Comment:
The `JobStreamerImpl` has tracks the liveness of the job stream via the `ZeebeFuture` returned when you send the command. It's important because we want to make sure that we recreate it in case it's unexpectedly terminated (except if the worker is closed, then we wouldn't). 

If we weren't locking whenever we access that, it could cause issues like having multiple streams open, or closing a stream we just opened, or keeping a stream open even after the worker is closed. This could happen since the different threads may complete the stream future, or access the streamer via the worker.